src/appleseed/renderer/modeling/object/curveobjectreader.cpp

```diff
@@ -75,369 +79,270 @@ namespace renderer
-        return object;
+        return auto_release_ptr<CurveObject>(nullptr);
     }
 
     const float radius = params.get_optional<float>("radius", 0.01f);
     if (radius <= 0.0f)
     {
         RENDERER_LOG_ERROR("curves radius must be greater than zero.");
-        return object;
+        return auto_release_ptr<CurveObject>(nullptr);
     }
 
-    vector<GVector3> vertices, new_vertices;
-
-    for (uint32 vertex_index = 0; vertex_index < vertex_count; )
-    {
-        float x;
-        if (fread(&x, sizeof(x), 1, file) != 1)
-        {
-            fclose(file);
-            RENDERER_LOG_ERROR("failed to load curve file %s: i/o error.", filepath.c_str());
-            return object;
-        }
+    GenericCurveFileReader reader(search_paths.qualify(filepath.c_str()).c_str(), radius, degree);
 
-        if (FP<float>::is_inf(x))
-        {
-            // End of the current hair strand, beginning of a new one.
+    CurveObjectBuilder builder(params, name);
 
-            switch (degree)
-            {
-              case 1:
-                if (vertices.size() >= 2)
-                {
-                    for (size_t i = 0, e = vertices.size() - 1; i < e; ++i)
-                    {
-                        const Curve1Type curve(&vertices[i], radius);
-                        object->push_curve1(curve);
-                    }
-                }
-                break;
-
-              case 3:
-                if (vertices.size() >= 4)
-                {
-                    assert(new_vertices.empty());
-
-                    for (size_t i = 0; i < vertices.size(); ++i)
-                    {
-                        new_vertices.push_back(vertices[i]);
-
-                        if (i > 0 && i % 2 == 0 && i + 1 < vertices.size())
-                        {
-                            // Add a midpoint.
-                            new_vertices.push_back(0.5f * (vertices[i] + vertices[i + 1]));
-                        }
-                    }
-
-                    for (size_t i = 0, e = new_vertices.size(); i + 3 < e; i += 3)
-                    {
-                        const Curve3Type curve(&new_vertices[i], radius);
-                        object->push_curve3(curve);
-                    }
-                }
-                break;
-
-              assert_otherwise;
-            }
-
-            clear_keep_memory(vertices);
-            clear_keep_memory(new_vertices);
+    Stopwatch<DefaultWallclockTimer> stopwatch;
+    stopwatch.start();
 
-            continue;
-        }
+    try
+    {
+        reader.read(builder);
+    }
 
-        Vector2f yz;
-        if (fread(&yz, sizeof(yz), 1, file) != 1)
-        {
-            fclose(file);
-            RENDERER_LOG_ERROR("failed to load curve file %s: i/o error.", filepath.c_str());
-            return object;
-        }
+    catch (const ExceptionIOError&)
+    {
+        RENDERER_LOG_ERROR(
+                "failed to load curve file %s: i/o error.",
```

Comment:
Make sure to set your IDE's indentation to 4 spaces (seems like you're using 8 here). Fix indentation everywhere.